### PR TITLE
Chore/resolve map and provider

### PR DESCRIPTION
**Pull request by Jakub-Vacek**
## Description

This PR refactors local map and provider resolution. It removes overly complicated `resolveValue` function and it introduces separate functions for local map and provider resolution. It also improves validation of loaded files.

This should NOT be braking change. But in cases of poorly configured SDK this change may lead to throwing valid errors (we are checking that eg. profile id in map ast matches profile id in super.json)

This PR is ready we just need to decide if this is shippable under minor change.

fixes: https://github.com/superfaceai/one-sdk-js/issues/279

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have updated the documentation accordingly. If you are changing **code related to user secrets** you need to really make sure that [security documentation](https://github.com/superfaceai/one-sdk-js/blob/main/SECURITY.md) is correct.
- [x] I have read the [CONTRIBUTING](https://github.com/superfaceai/one-sdk-js/blob/main/CONTRIBUTING.md) document.
- [x] I haven't repeated the code. (DRY)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/core/errors/errors.helpers.ts [hunk 1]
```diff
@@ -179,11 +179,24 @@ export function versionMismatchError(
...
 
+export function variantMismatchError(
+  astVariant: string | undefined,
+  passedVariant: string | undefined
+): SDKExecutionError {
+  return new SDKExecutionError(
+    `Variant from compiled map (${
+      astVariant ?? 'undefined'
+    }) and passed variant (${passedVariant ?? 'undefined'}) does not match.`,
+    ['If variant in map and passee variant is used they must match'],
```
Comment (martinalbert) [inline src/core/errors/errors.helpers.ts:195]:
```suggestion
    ['If variant in map and passed variant is used, they must match'],
```

File: src/core/errors/errors.helpers.ts [hunk 2]
```diff
@@ -192,6 +205,20 @@ export function profileNotFoundError(profileName: string): SDKExecutionError {
   );
 }
 
+export function profileProviderNotFoundError(
+  profileId: string,
+  providerName: string
+): SDKExecutionError {
+  return new SDKExecutionError(
+    `Provider "${providerName}" not found in profile "${profileId}" in super.json`,
```
Comment (martinalbert) [inline src/core/errors/errors.helpers.ts:213]:
```suggestion
    `Provider "${providerName}" not found for profile "${profileId}" in super.json`,
```

File: src/core/profile-provider/resolve-map-ast.test.ts [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+              },
+              defaults: {},
+              priority: [providerName],
+            },
+          },
+          providers: {},
+        },
+        config,
+      })
+    ).rejects.toThrow('File referenced in super.json not found');
```
Comment (martinalbert) [inline src/core/profile-provider/resolve-map-ast.test.ts:222]: Why not use error helper here?

File: src/core/profile-provider/resolve-map-ast.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        fileSystem.path.dirname(config.superfacePath),
+        profileProviderSettings.file.replace(
+          EXTENSIONS.map.source,
+          EXTENSIONS.map.build
+        )
+      );
+      // check if ast exists to print usefull error (needs to be compiled)
+      if (!(await fileSystem.exists(path))) {
+        throw sourceFileExtensionFoundError(EXTENSIONS.map.source);
+      }
```
Comment (martinalbert) [inline src/core/profile-provider/resolve-map-ast.ts:67]: question: Is this necessary when you're checking bellow whether reading file failed or not?

File: src/core/profile-provider/resolve-map-ast.test.ts [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+            },
+          },
+          providers: {},
+        },
+        config,
+      })
+    ).rejects.toThrow(providersDoNotMatchError('provider', 'different', 'map'));
+  });
+
+  it('return map ast when path to source is passed', async () => {
```
Comment (martinalbert) [inline src/core/profile-provider/resolve-map-ast.test.ts:361]: Can you also add here that it expects that ast is next to source, I thought that the next test is duplicate. 😄

File: src/core/provider/resolve-provider-json.test.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            [providerName]: {
+              file: 'path.json',
+              security: [],
+              parameters: {},
+            },
+          },
+        },
+        config,
+      })
+    ).rejects.toThrow('File referenced in super.json not found');
```
Comment (martinalbert) [inline src/core/provider/resolve-provider-json.test.ts:101]: Helpers not used here as well 🤷🏻‍♂️

**Review by martinalbert — APPROVED** [review]

File: src/core/profile-provider/resolve-map-ast.ts (hunk 4 above)
Comment (Jakub-Vacek) [reply src/core/profile-provider/resolve-map-ast.ts:67]: This is there to throw correct error when source file is not compiled.

**Comment by Jakub-Vacek** [comment]
@freaz @TheEdward162 what do you think about this PR. Are we able to merge it in this cycle?

**Review by TheEdward162 — APPROVED** [review]
> 4 months ago
aaaaaaaa :D

File: src/core/profile-provider/profile-provider.test.ts [hunk 7]
```diff
@@ -193,52 +147,54 @@ describe('profile provider', () => {
...
-              default: 'third-default',
-            },
-            {
-              name: 'fourth',
-              default: 'fourth-default',
-            },
-          ],
-        };
+        });
+        // jest.mocked(fetchBind).mockResolvedValue({
```
Comment (freaz) [inline src/core/profile-provider/profile-provider.test.ts:153]: leftover?

File: src/core/profile-provider/resolve-map-ast.test.ts (hunk 5 above)
Comment (freaz) [reply src/core/profile-provider/resolve-map-ast.test.ts:361]: Martin has a good point

Comment (freaz) [reply src/core/profile-provider/resolve-map-ast.test.ts:361]: nitpick: 
```suggestion
  it('returns map ast when path to source is passed', async () => {
```

**Review by freaz — APPROVED** [review]
Few nitpics, but looks good. And I believe it can be release as minor.

**Pull request merged**